### PR TITLE
Add "Extending API Platform" page

### LINKED ISSUE #866: Add a new "Extending API Platform" documentation entry
API Platform now provides good extension points that work for REST and GraphQL APIs the same time. The docs should be updated to explicitly discourage using custom controllers AND Symfony Kernel's events, and instead encourage to use the following extension points:

* [ ] Data Provider: Fetch data, custom computation, custom hydration
    * [ ] Alternative when using the built-in Doctrine data provider: use extensions and create custom Doctrine hydrators
* [ ] Symfony Denormalizer: to customize the object computed from the JSON payload
* [ ] Symfony Voter: to hook custom authorization logic
* [ ] Symfony Validation constraint: to hook custom validation logic
* [ ] Data Persister: to hook custom business logic or computation to trigger before, during or after persistence (ex: mail, call to an external API…)
* [ ] Symfony Normalizer: to customize the output (add field in the JSON, encode codes, dates…)
* [ ] Symfony Messenger integration: to create RPC, async, service-oriented endpoints, should be used in place of custom controllers (because the messenger integration is compatible with both REST and GraphQL, while custom controllers only work with REST)
* [ ] DTOs and data transformers: to create specialized representations of a (usually large) object or graph of objects also available through their own endpoints (for internal objects, a simple data provider and a public `@ApiResource` must be used instead). Using DTOs and data transformers should be rare. In most cases, using a class marked with `@ApiResource` representing the public data model exposed through the API (this class don't have to be mapped with a Doctrine entity, as explained here https://api-platform.com/docs/core/design/). It must not be tight to the internal data model. It's the responsibility of the Data Provider to fetch data from the persistence system and to convert it to a `@ApiResource` class. Data Transformers must not be used for this purpose.
* [ ] Symfony Kernel Event Listener: Customize the HTTP request or response (REST only, last resort solution)
* [ ]  TODO: use the error handler (new component) to hook custom error handling logic (see also https://github.com/api-platform/core/pull/2972)

### PR DESCRIPTION
**Pull request by thib92**
Resolves #866

### CODE REVIEW

File: core/extending.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+## Symfony Normalizer and Denormalizer
+
+As explained in [The Serialization Process](./serialization.md), API Platform uses [The Symfony Serializer Component](https://symfony.com/doc/current/components/serializer.html) extensively.
+
+If you need to customize the input or output objects on a global level in your API, you can customize the Serializer instance in the Symfony Container to hook custom logic to the process.
+
+## Symfony Validation Constraints
+
+If you need custom validation logic, you can have a look at [Symfony Validation Constraints](https://symfony.com/doc/current/reference/constraints.html).
```
Comment (dunglas) [inline core/extending.md]:
```suggestion
If you need custom validation logic, create  custom [Symfony Validation Constraints](https://symfony.com/doc/current/reference/constraints.html).
```

**Comment by dunglas** [comment]
Thanks @thib92! Do you mind if I push in your branch to add more sections?

**Comment by thib92** [comment]
@dunglas feel free to!

**Comment by dunglas** [comment]
Can you give me access to this branch (https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork)?

**Comment by thib92** [comment]
It should be allowed already!

**Comment by dunglas** [comment]
I added page that is just a summary table. But it would be nice to add examples, and to merge your tests into this? WDYT?

**Comment by thib92** [comment]
I'm not sure what we can add from this PR that you haven't already said.

Regarding examples, I'd guess they fit on individual pages right? And I think they all already have examples. How would you see this?

**Comment by GregoireHebert** [comment]
This PR can be closed.
https://api-platform.com/docs/core/extending/

**Pull request closed without merging**